qiita_db/handlers/tests/test_oauth2.py

```diff
@@ -0,0 +1,294 @@
+
+        obs = self.get('/qiita_db/artifacts/100/mapping/', headers={
+            'Authorization': 'WRONG ' + self.client_token})
+        self.assertEqual(obs.code, 400)
+        exp = {'error': 'invalid_grant',
+               'error_description': 'Oauth2 error: invalid access token'}
+        self.assertEqual(loads(obs.body), exp)
+
+
+class OAuth2HandlerTests(TestHandlerBase):
+    def test_authenticate_client(self):
+        # Authenticate using header
+        obs = self.post(
+            '/qiita_db/authenticate/', {'grant_type': 'client'}, {
+                'Authorization': 'Basic MTluZGtPM29NS3NvQ2hqVlZXbHVGN1FreEhSZl'
+                                 'loVEtTRmJBVnQ4SWhLN2daZ0RhTzQ6SjdGZlE3Q1FkT3'
+                                 'h1S2hRQWYxZW9HZ0JBRTgxTnM4R3UzRUthV0ZtM0lPMk'
+                                 'pLaEFtbUNXWnVhYmUwTzVNcDI4czE='})
+        self.assertEqual(obs.code, 200)
+        obs_body = loads(obs.body)
+        exp = {'access_token': 'token',
+               'token_type': 'Bearer',
+               'expires_in': '3600'}
+        self.assertItemsEqual(obs_body.keys(), exp.keys())
+        self.assertEqual(obs_body['token_type'], exp['token_type'])
+        self.assertEqual(obs_body['expires_in'], exp['expires_in'])
+        self.assertEqual(len(obs_body['access_token']), 55)
+        self.assertEqual(type(obs_body['access_token']), unicode)
+
+        # Make sure token in system with proper ttl
+        token = r_client.hgetall(obs_body['access_token'])
+        self.assertNotEqual(token, {})
+        self.assertItemsEqual(token.keys(), ['timestamp', 'client_id'])
+        self.assertEqual(r_client.ttl(obs_body['access_token']), 3600)
+
+        # Authenticate using post only
+        obs = self.post(
+            '/qiita_db/authenticate/', {
+                'grant_type': 'client',
+                'client_id': '19ndkO3oMKsoChjVVWluF7QkxHRfYhTKSFbAVt8IhK7gZgDa'
+                             'O4',
+                'client_secret': 'J7FfQ7CQdOxuKhQAf1eoGgBAE81Ns8Gu3EKaWFm3IO2J'
+                                 'KhAmmCWZuabe0O5Mp28s1'})
+        self.assertEqual(obs.code, 200)
+        obs_body = loads(obs.body)
+        exp = {'access_token': 'placeholder',
+               'token_type': 'Bearer',
+               'expires_in': '3600'}
+        self.assertItemsEqual(obs_body.keys(), exp.keys())
+        self.assertEqual(obs_body['token_type'], exp['token_type'])
+        self.assertEqual(obs_body['expires_in'], exp['expires_in'])
+        self.assertEqual(len(obs_body['access_token']), 55)
+        self.assertEqual(type(obs_body['access_token']), unicode)
+
+        # Make sure token in system with proper ttl
+        token = r_client.hgetall(obs_body['access_token'])
+        self.assertNotEqual(token, {})
+        self.assertItemsEqual(token.keys(), ['timestamp', 'client_id'])
+        self.assertEqual(r_client.ttl(obs_body['access_token']), 3600)
+
+    def test_authenticate_client_bad_info(self):
```

Comment:
Split this in multiple tests - please see my email from Oct 20th about using AAA for tests.